File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque
answer = 0
m,n = map(int, (input().split(" ")))
graph = []
   # 상하좌우
dx = [-1,1,0,0]
dy = [0,0,-1,1]
queue = deque()
for i in range(n):
    graph.append(list(map(int, (input().split()))))
    for j in range(m):
        if graph[i][j] == 1:
            queue.append((i,j))
def bfs():
    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx = dx[i]+x
            ny = dy[i]+y
            if 0<= nx < n and 0<=ny<m and graph[nx][ny] == 0:
                graph[nx][ny] = graph[x][y] + 1
                queue.append((nx,ny))
bfs()
tomato = True
#토마토가 안자란곳이 한곳이라도 있으면 False

for i in graph:
    for j in i:
        if j==0:
            print(-1)
            exit(0)
    answer = max(answer,max(i))
print(answer-1)

File: 백준/Gold/7576. 토마토/토마토.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

class Point {
    int x;
    int y;

    Point(int x, int y) {
        this.x = x; //세로
        this.y = y; //가로
    }
}

public class Main {
    static StringTokenizer st;
    static int[][] data;
    static Queue<Point> tomato = new LinkedList<>();
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int day = 0;
    static int m, n;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        //가로
        m = Integer.parseInt(st.nextToken());
        //세로
        n = Integer.parseInt(st.nextToken());
        data = new int[n][m];
        //맵 생성
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                data[i][j] = Integer.parseInt(st.nextToken());
                if (data[i][j] == 1) {
                    tomato.add(new Point(i, j));
                }
            }
        }

        System.out.println(BFS());


    }

    static int BFS() {
        while (!tomato.isEmpty()) {

            day += 1;
            Point point = tomato.poll();
            int x = point.x;
            int y = point.y;
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
                    if (data[nx][ny] == 0) {
                        data[nx][ny] = data[x][y] + 1;
                        tomato.add(new Point(nx, ny));
                    }
                }
            }
        }

        int day = Integer.MIN_VALUE;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (data[i][j] == 0) {
                    return -1;
                }
                day = Math.max(day, data[i][j]);
            }
        }
        return day - 1;
    }
}
